pyfr/solvers/euler/inters.py

```diff
@@ -124,3 +129,120 @@ def __init__(self, be, lhs, elemap, cfgsect, cfg):
+    def __init__(self, be, lhs, elemap, cfgsect, cfg, bccomm):
+        super().__init__(be, lhs, elemap, cfgsect, cfg, bccomm)
+
+        self.c |= self._exp_opts(
+            ['rho', 'u', 'v', 'w'][:self.ndims + 1], lhs
+        )
+
+        self.target_mfr = self.cfg.getfloat(cfgsect, 'mass-flow-rate')
+        # When to start the mass flow controller
+        self.tstart = self.cfg.getfloat(cfgsect, 'tstart', 0.0)
+        # Start p value
+        self.p = self.cfg.getfloat(cfgsect, 'p')
+        self.bcname = cfgsect.removeprefix('soln-bcs-')
+        # Mass flow average
+        self.mf_avg = 0.0
+        self.mf_alpha = 0.4
+        # Parameter to control the strength of the controller
+        self.eta = self.cfg.getfloat(cfgsect, 'eta')
+        # Frequency that mf.csv should be updated
+        self.nsteps = self.cfg.getint(cfgsect, 'nsteps', 100)
+        self.nflush = self.cfg.getint(cfgsect, 'nflush', 10)
+
+        self._set_external('var_p', 'scalar fpdtype_t')
+        self.tprev = -1.0
+        self.nstep_counter = 0
+        self.nflush_counter = 0
+        
+        self.mf_int = MassFlowIntegrator(cfg, cfgsect, elemap, lhs)
+
+        if self.bccomm.rank == 0:
+            self.outf = init_csv(self.cfg, self.cfgsect, 't,mf,pbc')
+
+    def calculate_mass_flow(self, solns):
+        ndims, nvars = self.ndims, self.nvars
+        fm = np.zeros((ndims))
+        # Get the sizes for the area calculation
+        for etype, fidx in self.mf_int.m0:
+            # Get the interpolation operator
+            m0 = self.mf_int.m0[etype, fidx]
+            nfpts, nupts = m0.shape
+
+            # Extract the relevant elements from the solution
+            uupts = solns[etype][..., self.mf_int.eidxs[etype, fidx]]
+
+            # Interpolate to the face
+            ufpts = m0 @ uupts.reshape(nupts, -1)
+            ufpts = ufpts.reshape(nfpts, nvars, -1)
+            ufpts = ufpts.swapaxes(0, 1)
+
+            # Get the quadrature weights and normal vectors
+            qwts = self.mf_int.qwts[etype, fidx]
+            norms = self.mf_int.norms[etype, fidx]
+
+            # Do the quadrature for each dimension
+            # RhoU = ufpts[1], RhoV = ufpts[2], RhoW = ufpts[2]
+            for i in range(0, ndims):
+                rhoVel = ufpts[1 + i]
+                fm[i] += np.einsum('i...,ij,ji', qwts, rhoVel, norms[:,:,i])
+        self.bccomm.Allreduce(mpi.IN_PLACE, fm, op=mpi.SUM)
+        return sum(fm)
+    
+    def update_mf(self, solns):
+        mf = self.calculate_mass_flow(solns)
+        self.mf_avg = self.mf_alpha * mf + (1.0 - self.mf_alpha) * self.mf_avg
+
+    def update_p(self, dt):
+        self.p += dt * self.eta * (1.0 - self.target_mfr / self.mf_avg)
+    
+    def bind_p(self, kerns):
+        for k in kerns:
+            kerns[k].bind(var_p=self.p)
+
+    def prepare(self, system, ubank, t, kerns):
+        # Check if past tstart
+        if t < self.tstart:
+            self.bind_p(kerns)
```

Comment:
See if you can refactor the logic so we don't have bind_p called in three different places.